src/packed-pack.ts

```diff
@@ -283,19 +280,101 @@ export class PackContent {
                 reader.offset = relOffset;
             }
 
-            // n-byte type and length (3-bit type, (n-1)*7+4-bit length) compressed data
-            // https://git-scm.com/docs/pack-format#_size_encoding
-            const int = readVarIntLE(reader);
-            const btype = ((int >> 4) & 0b0111) as PackedType;
-            const length = ((int >> 7) << 4) | (int & 0b1111);
+            const objectPrelude = await this.readObjectPrelude(offsetBase, reader);
+            const { btype, unpackedSize } = objectPrelude;
 
             objectsByType[btype].count++;
             objectsByType[btype].size += nextOffset - offset;
-            objectsByType[btype].unpackedSize += length;
+            objectsByType[btype].unpackedSize += unpackedSize;
+
+            if (isOffsetDeltaPrelude(objectPrelude)) {
+                objectsByType[btype].unpackedRestoredSize += objectPrelude.unpackedRestoredSize;
+            }
         }
 
         return objectsByType.filter((stat) => stat.count > 0);
     }
+
+    private read(buffer: Buffer, offset: number) {
+        return this.fh.read(buffer, 0, buffer.byteLength, offset);
+    }
+
+    private async readObjectPreludeFromFile(
+        offset: number
+    ): Promise<[BufferCursor, ObjectPrelude]> {
+        const header =
+            reuseBufferCount > 0 ? buffers[--reuseBufferCount] : Buffer.allocUnsafe(4096);
+
+        await this.read(header, offset);
+        const reader = new BufferCursor(header);
+
+        return [reader, await this.readObjectPrelude(offset, reader)];
+    }
+
+    private async readObjectPrelude(
+        offsetBase: number,
+        reader: BufferCursor
+    ): Promise<ObjectPrelude> {
+        // object start relative to current buffer
+        const relOffset = reader.offset;
+
+        // https://git-scm.com/docs/pack-format#_size_encoding
+        // n-byte type and length (3-bit type, (n-1)*7+4-bit length) compressed data
+        const int = readVarIntLE(reader);
+        const btype = ((int >> 4) & 0b0111) as PackedType;
+        const length = ((int >> 7) << 4) | (int & 0b1111);
+
+        if (btype === OBJ_INVALID || btype === OBJ_RESERVED) {
+            throw new Error(`Unrecognized type: 0b${btype.toString(2)}`);
+        }
+
+        const result = {
```

Comment:
Using different shapes for a result is not good for performance (turns functions in polymorphic state). It's better to keep the same set of fields for a result even some fields don't used.